### PR TITLE
Add DiscreteFunction

### PR DESCRIPTION
**Pull request by ErikQQY**
Part of #993

### CODE REVIEW

File: test/discretesystem.jl [hunk 1]
```diff
@@ -28,6 +28,7 @@ eqs = [D(S) ~ S - infection * h,
 @named sys = DiscreteSystem(eqs, t, [S, I, R], [c, nsteps, δt, β, γ]; controls = [β, γ])
 syss = structural_simplify(sys)
 @test syss == syss
+df = NonlinearFunction(sys)
```
Comment (ChrisRackauckas) [inline test/discretesystem.jl:31]:
```suggestion
df = DiscreteFunction(sys)
```

**Review by ChrisRackauckas — APPROVED** [review]

**Review by YingboMa — CHANGES_REQUESTED** [review]
Needs formatting and tests. See https://github.com/SciML/ModelingToolkit.jl/blob/master/CONTRIBUTING.md

**Comment by ErikQQY** [comment]
The out-of-place tests are fine, but I still can't figure out what is wrong with the inplace discrete functions.

image

**Comment by YingboMa** [comment]
Isn't `f(u, p, t)` out-of-place? You can inspect the function body by printing `f.body` if `f` is a RuntimeGeneratedFunction.

**Comment by ErikQQY** [comment]
Yeah, made a mistake, ```f(u, p, t)``` is out-of-place🥲.

Just fixed all of the tests!

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
@YingboMa what's going on with the docs failure?

**Comment by YingboMa** [comment]
I will take a look on Monday.

**Pull request merged**